File: 1163-last-substring-in-lexicographical-order/1163-last-substring-in-lexicographical-order.cpp
class Solution {
public:
    string lastSubstring(string s) {
        char ma = 'A';
        int midx = s.size() - 1;
        for (int i = s.size()-1; i >= 0; i--) {
            if (ma < s[i]) {
                midx = i;
                ma = s[i];
            }
            else if (ma == s[i]) {
                int j = i+1,
                    jm = midx + 1;
                while (j < midx && jm < s.size()) {
                    if (s[j] < s[jm])
                        break;
                    else if (s[j] > s[jm])
                        j = midx-1;
                    j++;
                    jm++;
                }
                if (j == midx || jm == s.size())
                    midx = i;
            }
        }
        return s.substr(midx);
    }
};

File: 1163-last-substring-in-lexicographical-order/1163-last-substring-in-lexicographical-order.py
class Solution:
    def lastSubstring(self, s: str) -> str:
        ma = 'A'
        midx = len(s)-1
        for i in range(len(s)-1, -1, -1):
            if ma < s[i]:
                midx = i
                ma = s[i]
            elif ma == s[i]:
                j = i+1
                jm = midx + 1
                while j < midx and jm < len(s):
                    if s[j] < s[jm]:
                        break
                    elif s[j] > s[jm]:
                        j = midx-1
                    j += 1
                    jm += 1
                if j == midx or jm == len(s):
                    midx = i
        return s[midx:]
            